Initial implementation of InsertObserverNodes pass.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18152 Initial implementation of InsertObserverNodes pass.**
* #18151 Add quant-passes stubs.

gh-metadata: pytorch pytorch 18150 gh/zolotukhinm@gmail.com/2/head

Differential Revision: [D14584223](https://our.internmc.facebook.com/intern/diff/D14584223)

Commit:
Initial implementation of InsertObserverNodes pass.

gh-metadata: pytorch pytorch 18150 gh/zolotukhinm@gmail.com/2/head

apaszke commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -18,8 +18,30 @@ void PropagateQuantInfo(std::shared_ptr<Graph>& graph) {
-  throw std::runtime_error("Pass not implemented yet!");
+  for (const auto& n : graph->nodes()) {
+    // Skip nodes that we've just added
+    if (n->kind().toQualString() == std::string("aten::observer")) {
```
What's `aten::observer`? A grep shows no hits in our codebase

ZolotukhinM replied:
You are correct, it doesn't exist yet.

This diff is a part of an effort to enable quantization. The given pass would insert so called observer nodes, which collect distribution of values taken by a given tensor. This data then will be used for computing scale and bias of quantization. The work on the observer operator itself is carried on in parallel (the name `aten::observer` is tentative).

Commit:
Update on "Initial implementation of InsertObserverNodes pass."

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -18,8 +18,30 @@ void PropagateQuantInfo(std::shared_ptr<Graph>& graph) {
+  observer->moveAfter(def);
+  v->replaceAllUsesWith(observer->outputs()[0]);
+  observer->outputs()[0]->setUniqueName(v->uniqueName() + ".observed");
+  observer->addInput(v);
```
Comment: input of v is added here because adding it at node creation would make it a use of v and subject to replaceAllUsesWith

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -18,8 +18,30 @@ void PropagateQuantInfo(std::shared_ptr<Graph>& graph) {
+  torch::jit::Node* observer =
+      def->owningGraph()->create(at::Symbol::fromQualString("aten::observer"));
+  def->owningBlock()->appendNode(observer);
+  observer->moveAfter(def);
```
No need to append and then move. It can just be `observer->insertAfter`

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -18,8 +18,30 @@ void PropagateQuantInfo(std::shared_ptr<Graph>& graph) {
-  throw std::runtime_error("Pass not implemented yet!");
+  for (const auto& n : graph->nodes()) {
+    // Skip nodes that we've just added
+    if (n->kind().toQualString() == std::string("aten::observer")) {
```
can you add aten::observer to the interned_strings.h header so we don't reparse it everytime?

ZolotukhinM replied:
I would leave that till we actually implement the observer op. E.g. I don't know for sure at the moment whether it will be `aten::` or something else.

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -18,8 +18,30 @@ void PropagateQuantInfo(std::shared_ptr<Graph>& graph) {
+    for (torch::jit::Value* v : n->outputs()) {
+      addObserverFor(v);
+    }
+  }
```
Are you intending to skip all subblocks?

ZolotukhinM replied:
We will need to visit subblocks as well, good catch. I'll update the PR accordingly (I'll also need to refine its logic to e.g. work correctly with `prim::` nodes).

zdevito reviewed (approved):
Looks good. Do you want to see all nodes, or just the top-level blocks nodes?

ZolotukhinM commented:
Thanks for reviewing! I'll update the patch shortly.

Commit:
Update on "Initial implementation of InsertObserverNodes pass."

Initial implementation of InsertObserverNodes pass.

gh-metadata: pytorch pytorch 18152 gh/ZolotukhinM/2/head

ZolotukhinM commented:
@zdevito, could you please take another look if you have time? The PR has changed quite significantly since you approved it.

dzhulgakov commented on torch/csrc/jit/init.cpp:
```diff
@@ -122,7 +122,12 @@ void initJITBindings(PyObject* module) {
           "_jit_pass_insert_observers",
-          [](std::shared_ptr<Graph>& g) { return InsertObserverNodes(g); })
+          [](std::shared_ptr<Graph>& g, py::function pp) {
+            Node* new_node =
```
ok, so you use "name" as identity, that works too. Another option I was playing with is to pass into a "class" instance and create new instance for each observer not we insert. This way we don't need to rely on any names at all (other for debugging): https://github.com/pytorch/pytorch/compare/master...dzhulgakov:observer-hack-illustration?expand=1 @nishantpdce

ZolotukhinM replied:
Right. I think both options are almost equivalent. The advantage of this approach would be:
- Potential better debugging experience, as you mentioned,
- It's more transparent to the user,
- Less coupling between passes: I can remove observers and save the model, then I can load it and immediately quantize. Alternatively, I can quantize even without running the observers if I prepare the data myself in the correct format,
- It's already done here :)

PS: Also, thanks for commenting on exactly this line of code - I noticed some pieces that I forgot to cleanup!

dzhulgakov replied:
Regarding loading - do we guarantee preservation of unique_name through serialization?

The question is whether it's safe to rely on uniqueName being stable between passes? The benefit of explicit objects is that it'd preserve identity of an operator explicitly, even with nested blocks/modules. E.g. imagine that you're going to start quantizing a set of several nested modules that can call each other. You'd need to start passing method and module names as part of the fully qualified name too. Or we'd be quantizing on per function basis only?

In general I agree they are equivalent at this point.

ZolotukhinM replied:
There are no guarantees regarding the names, but it shouldn't be an issue since we embed the actual name as a constant parameter into every call to the observer function. Thus, when we use the dictionary, the key would be that second argument, not the unique name of the value. In practice, they would most probably match, but there are no guarantees on that.

As for the nested submodules, I expect we might need to refine the way we construct the key in the dictionary (e.g. use "scope:unique_name" instead of just "unique_name"), but overall the approach should still work.

Note btw that this approach can be used for inserting any type of observers, not just for quantization observers. E.g. if the observer function records timing since its last invocation, you would get a time profile for free.

raghuramank100 replied:
Are names preserved across passes? For example, if we have passes between observer and quantizer, can we assume that the same names exist?

nishantpdce replied:
It would be good to run this through with simple LeNet (script and trace) just to make sure all nodes are visited and get the observer

ZolotukhinM replied:
Names are note guaranteed to be preserved across passes, however this approach would still work correctly because we embed the names into the observer calls.

The correspondence between the old observed stats and a value would be restored from the arguments of the observer call and will not depend on the name of the value. E.g. in
`x1 = observe(x, "some_name")
`
We know that `value_stats["some_name"]` ("some_name" is the 2nd argument) contains observed stats for `x` (`x` is the 1st argument).

Note that we can use any other identifier instead of the name here, and it would work just as well, using the name just gives us some inspectability for free.

Commit:
Update on "Initial implementation of InsertObserverNodes pass."

Initial implementation of InsertObserverNodes pass.

gh-metadata: pytorch pytorch 18152 gh/ZolotukhinM/2/head

Commit:
Update on "Initial implementation of InsertObserverNodes pass."

Initial implementation of InsertObserverNodes pass.

gh-metadata: pytorch pytorch 18152 gh/ZolotukhinM/2/head

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -18,8 +17,74 @@ void PropagateQuantInfo(std::shared_ptr<Graph>& graph) {
 
-void InsertObserverNodes(std::shared_ptr<Graph>& graph) {
-  throw std::runtime_error("Pass not implemented yet!");
+static void addObserverFor(Value* v, Node* original_observer_node) {
```
can we keep all the quantizer util methods at the top
